src/main/java/karen/Parser.java

```diff
@@ -0,0 +1,223 @@
+import karen.command.ModifyType;
+import karen.task.Deadline;
+import karen.task.Event;
+import karen.task.ToDo;
+
+/**
+ * Validates and parses arguments to create Command objects.
+ */
+public class Parser {
+    public static final Pattern INDEX_FORMAT = Pattern.compile("^(?<keyWord>[^\\s]+)\\s+(?<index>\\d+)$");
+    public static final Pattern KEYWORD_FORMAT = Pattern.compile("^(?<keyWord>[^\\s]+).*$");
+    public static final Pattern TODO_FORMAT = Pattern.compile("^todo\\s+(?<description>.*)$");
+    public static final Pattern DEADLINE_FORMAT = Pattern.compile("^deadline (?<description>.*) \\/by (?<time>.*)$");
+    public static final Pattern EVENT_FORMAT = Pattern.compile("^event (?<description>.*) \\/at (?<time>.*)$");
+    public static final Pattern FIND_FORMAT = Pattern.compile("^find (?<keyTerm>.*)$");
+
+    public static final String NA_MESSAGE = "I don't understand anything - I want to speak with your manager";
+
+    /**
+     * Validates if dateString parameter follows yyyy-mm-dd format.
+     * @param dateString input string
+     * @return dateString original input string
+     * @throws KarenException if format doesn't follow yyyy-mm-dd
+     */
+    public String validateDateFormat(String dateString) throws KarenException {
+        try {
+            LocalDate.parse(dateString);
+        } catch (DateTimeException err) {
+            throw new KarenException("Wrong date formatting. It should be in yyyy-mm-dd");
+        }
+        return dateString;
+    }
+
+    /**
+     * Creates InvalidCommand objects with troubleshooting messages
+     * @param keyWord first word of input
+     * @param fullInput full input from user
+     * @return InvalidCommand with default or custom message.
+     */
+    private Command prepareInvalid(String keyWord, String fullInput) {
+        Command cmd = new InvalidCommand();
+
+        switch (keyWord) {
+        case "deadline":
+            if (fullInput.matches("^((?!\\/by).)*$")) {
+                cmd = new InvalidCommand("You're missing an /by flag needed to add deadlines");
+            }
+            break;
+        case "event":
+            if (fullInput.matches("^((?!\\/at).)*$")) {
+                cmd = new InvalidCommand("You're missing an /at flag needed to add events");
+            }
+        }
+        return cmd;
+    }
+
+    /**
+     * Creates AddCommand object by parsing parameters to instantiate Task objects and
+     * other relevant arguments.
+     * @param keyWord first word of input
+     * @param fullInput full input from user
+     * @return AddCommand object
+     */
+    private Command prepareAdd(String keyWord, String fullInput) {
+        final Matcher matcher;
+        try {
+            if (keyWord.equals("todo")) {
+                matcher = TODO_FORMAT.matcher(fullInput);
+                matcher.find();
+                return new AddCommand(new ToDo(matcher.group("description")));
+            } else if (keyWord.equals("deadline")) {
```

Comment:
Maybe you can consider using switch case here so that it is easier to manage when more keywords might be added in the future.